docs/src/index.md

```diff
@@ -3,8 +3,120 @@
 ```@index
 ```
 
+## Interface
+
 ```@docs
 ConstructionBase
 ConstructionBase.constructorof
 ConstructionBase.setproperties
 ```
+
+## [Tips for designing types](@id type-tips)
+
+When designing types from scratch, it is recommended to structure the types in
+such a way that overloading `constructorof` or `setproperties` is unnecessary in
+the first place.  For simple `struct`s whose type parameters can be determined
+from field values, `ConstructionBase` works without any customization, provided
+that the "type-less" constructor exists.  However, it is often useful or
+required to have type parameters that cannot be determined from field values.
+One way to solve this problem is to define singleton types that would determine
+the type parameters:
+
+```jldoctest tips
+abstract type OutputBy end
+struct Mutating <: OutputBy end
+struct Returning <: OutputBy end
+
+struct Add{O <: OutputBy, T}
+    outputby::O
+    value::T
+end
+
```

Comment:
I like this trick very much and it did not occur to me before. I also like that it lets you manipulate the `O` via `setproperties`.

Reply:
I guess I was thinking backward:  Given Setfield, how can I make my types better? :smile: